docs/sources/userguide/plugins.md

```diff
@@ -0,0 +1,157 @@
+
+```
+$ docker plugin load -e $ARGS clusterhq/flocker-plugin $@
+[... fetching...]
+Plugin flocker:v0.1 loaded and registered with extension points:
+* volumes
+Plugin flocker activated.
+```
+
+This is really just syntactic sugar for the following:
+
+```
+$ docker run -d -e $ARGS \
+	-v /var/run/docker-plugins/flocker.sock:/var/run/plugin.sock \
+	-v /var/run/docker.sock:/var/run/docker.sock \
+	clusterhq/flocker-plugin $@
+```
+
+(The docker socket should only be mounted if the plugin is started with `--privileged`.)
+
+Loading a plugin forces it to always be loaded when Docker restarts (and Docker doesn't respond to API requests until it completes loading all its plugins).
+
+Docker then waits for the plugin to start listening on the socket (it polls the socket until it gets a successful response to an HTTP query to `/v1/plugin/handshake` on the socket, which returns with just a list of subsystems the plugin is interested in - response defined below).
+According to the type of plugin which is negotiated in the handshake, Docker registers the plugin to send it HTTP requests on certain events.
+
+Plugins should name themselves in the response, which should be in this format:
+
+```
+{InterestedIn: ["volume"],
+ PluginName: "flocker"}
+```
+
+Every event should be sent to every plugin registered for that subsystem for now.
+Later the handshake can include more granular negotiation over which events it wants to receive.
+
+Other `plugin` subcommands can include `list`, `status`, `upgrade`, and `unload`.
+
+# Types of plugin
+
+Each plugin type defines a straightforward (MVP) protocol.
+
+## Volume
+
+We will start with this first because it is has the tiniest interface.
+
+The simplest of plugin types, `volume` provides a single request-response type on `POST /v1/volumes` (`POST ~= create`):
+
+**Request**
+
+```
+{HostPath: "/path",
+ ContainerID: "abcdef123"}
+```
+
+**Response**
+
+```
+{HostPath: "/newpath",
+ ContainerID: "abcdef123"}
+```
+
+In the initial version, if the plugin responds with a 404 then the response is ignored.
+
+See reference implementation:
+
+* [Docker volumes extension mechanism](https://github.com/cpuguy83/docker/compare/ddb366ee9a07e3feab766cc712c9683ad0c3c309)
+* [Flocker extension](https://github.com/clusterhq/powerstrip-flocker/compare/docker-volume-extension)
+
+**Demo**
+
+Despite the name, this demo does not use powerstrip!
+
+```
+$ git clone https://github.com/clusterhq/powerstrip-flocker
+$ cd powerstrip-flocker/vagrant-aws
+$ git checkout docker-volume-extension
+$ vagrant up --provider=aws
+$ vagrant ssh node1
+node1$ docker run -v /flocker/test:/data ubuntu sh -c "echo fish > /data/file.txt"
+node1$ exit
+$ vagrant ssh node2
+node2$ docker run -v /flocker/test:/data ubuntu cat /data/file.txt
+fish
+```
+
+In future, we will want to extend this to notify volume extensions when containers start and stop, so that the underlying volumes mechanism can hold and release *leases* on volumes.
```

Comment:
If an extension has access to the docker socket - it can listen to the event stream to know if a container has stopped.  Is it better to leave this up to the extension?  The same issue would be true of networking extensions that want to release an IP.